File: rxbilling-google-play-library-v4/src/main/java/com/vanniktech/rxbilling/google/play/library/v4/PlayBillingInventorySubscription.java
package com.vanniktech.rxbilling.google.play.library.v4;

import androidx.annotation.NonNull;
import com.android.billingclient.api.SkuDetails;
import com.google.auto.value.AutoValue;
import com.google.auto.value.extension.memoized.Memoized;
import com.vanniktech.rxbilling.InventorySubscription;
import java.math.BigDecimal;
import java.util.Currency;

@SuppressWarnings("PMD.UseObjectForClearerAPI") @AutoValue abstract class PlayBillingInventorySubscription implements InventorySubscription {
  @NonNull public static InventorySubscription create(@NonNull final SkuDetails skuDetails) {
    return new AutoValue_PlayBillingInventorySubscription(skuDetails.getSku(), skuDetails.getType(), skuDetails.getPrice(), skuDetails.getPriceAmountMicros(), skuDetails.getPriceCurrencyCode(),
        skuDetails.getTitle(), skuDetails.getDescription(), skuDetails);
  }

  @Override @NonNull @Memoized public Currency priceCurrency() {
    return Currency.getInstance(priceCurrencyCode());
  }

  @Override @NonNull @Memoized public BigDecimal priceAsBigDecimal() {
    return Utils.microsAsBigDecimal(priceAmountMicros());
  }

  @NonNull abstract SkuDetails skuDetails();
}

File: rxbilling-google-play-library-v5/src/main/kotlin/com/vanniktech/rxbilling/google/play/library/v5/PlayBillingInventorySubscription.kt
@file:JvmName("Utils")

package com.vanniktech.rxbilling.google.play.library.v5

import com.android.billingclient.api.ProductDetails
import com.android.billingclient.api.ProductDetails.PricingPhase
import com.vanniktech.rxbilling.InventorySubscription
import java.math.BigDecimal
import java.util.Currency

data class PlayBillingInventorySubscription(
  val productDetails: ProductDetails,
  val pricingPhase: PricingPhase,
  val offerToken: String,
  val offerTags: List<String>,
) : InventorySubscription {
  override val sku: String = productDetails.productId
  override val price: String = pricingPhase.formattedPrice
  override val priceAmountMicros: Long = pricingPhase.priceAmountMicros
  override val priceCurrencyCode: String = pricingPhase.priceCurrencyCode
  override val type: String = productDetails.productType
  override val title: String = productDetails.title
  override val description: String = productDetails.description

  override fun priceCurrency(): Currency = Currency.getInstance(priceCurrencyCode)
  override fun priceAsBigDecimal(): BigDecimal = priceAmountMicros.microsAsBigDecimal()
}
